snapshot/v3_snapshot.go

```diff
@@ -0,0 +1,456 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package snapshot
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/binary"
+	"encoding/json"
+	"fmt"
+	"hash/crc32"
+	"io"
+	"math"
+	"os"
+	"path/filepath"
+	"reflect"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/coreos/etcd/etcdserver"
+	"github.com/coreos/etcd/etcdserver/etcdserverpb"
+	"github.com/coreos/etcd/etcdserver/membership"
+	"github.com/coreos/etcd/lease"
+	"github.com/coreos/etcd/mvcc"
+	"github.com/coreos/etcd/mvcc/backend"
+	"github.com/coreos/etcd/pkg/fileutil"
+	"github.com/coreos/etcd/pkg/types"
+	"github.com/coreos/etcd/raft"
+	"github.com/coreos/etcd/raft/raftpb"
+	"github.com/coreos/etcd/snap"
+	"github.com/coreos/etcd/store"
+	"github.com/coreos/etcd/wal"
+	"github.com/coreos/etcd/wal/walpb"
+
+	bolt "github.com/coreos/bbolt"
+	"google.golang.org/grpc/grpclog"
+)
+
+// Snapshotter defines snapshot methods.
+type Snapshotter interface {
+	// Save fetches snapshot from specified client's endpoints and saves to target path.
+	// If the context "ctx" is canceled or timed out, snapshot save stream will error out
+	// (e.g. context.Canceled, context.DeadlineExceeded).
+	Save(ctx context.Context, dbPath string) error
+
+	// Status returns the snapshot file information.
+	Status(dbPath string) (DBStatus, error)
+
+	// Restore restores a new etcd data directory from given snapshot file.
+	Restore(dbPath string, cfg RestoreConfig) error
+}
+
+// DBStatus is the snapshot file status.
+type DBStatus struct {
+	Hash      uint32 `json:"hash"`
+	Revision  int64  `json:"revision"`
+	TotalKey  int    `json:"totalKey"`
+	TotalSize int64  `json:"totalSize"`
+}
+
+// RestoreConfig configures snapshot restore operation.
+type RestoreConfig struct {
+	// Name is the human-readable name of this member.
+	Name string
+	// OutputDataDir is the target data directory to save restored data.
+	// Defaults to "[Name].etcd" if not given.
+	OutputDataDir string
+	// OutputWALDir is the target WAL data directory.
+	// Defaults to "[OutputDataDir]/member/wal" if not given.
+	OutputWALDir string
+	// InitialCluster is the initial cluster configuration for restore bootstrap.
+	InitialCluster types.URLsMap
+	// InitialClusterToken is the initial cluster token for etcd cluster during restore bootstrap.
+	InitialClusterToken string
+	// PeerURLs is a list of member's peer URLs to advertise to the rest of the cluster.
+	PeerURLs types.URLs
+	// SkipHashCheck is "true" to ignore snapshot integrity hash value
+	// (required if copied from data directory).
+	SkipHashCheck bool
+}
+
+// Config defines snapshot config.
+type Config struct {
```

Comment:
we probably should kill this config struct. this is not snapshot related configuration.